feature/641

sam-irl opened the pull request:
Resolves #641 

The purpose of this pull request is to add commands to say an aircraft's (assigned) speed, altitude, and heading

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,78 @@ export default class AircraftCommander {
+    runSayAltitude(aircraft) {
+        const altitude = aircraft.altitude;
+
+        return [true, `our altitude is ${altitude}`];
```
```
const altitude = _round(aircraft.altitude, -2);
const isClimbingOrDescending = aircraft.trend !== 0;
let altitudeChangeString = '';

if (isClimbingOrDescending) {
    altitudeChangeString = 'leaving ';
}

return [true, `${altitudeChangeString}${altitude}];`
```

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,78 @@ export default class AircraftCommander {
+    runSayAssignedAltitude(aircraft) {
+        const altitude = aircraft.target.altitude;
+
+        return [true, `assigned altitude ${altitude}`];
```
```
const altitude = _round(aircraft.mcp.altitude, -2);

return [true, `assigned ${altitude}];`
```

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,78 @@ export default class AircraftCommander {
+    runSayHeading(aircraft) {
+        const heading = aircraft.heading;
+
+        return [true, `our heading is ${heading}`];
```
```
const heading = _round(radiansToDegrees(aircraft.heading));

return [true, `heading ${heading}`];
```

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,78 @@ export default class AircraftCommander {
+    runSayAssignedHeading(aircraft) {
+        const heading = aircraft.target.heading;
+
+        return [true, `assigned heading ${heading}`];
```
```
if (aircraft.mcp.headingMode !== MCP_MODE.HEADING.HOLD) {
    return [true, 'we haven't been assigned a heading'];
}

const heading = _round(radiansToDegrees(aircraft.mcp.heading));

return [true, `assigned ${heading} heading`];
```

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,78 @@ export default class AircraftCommander {
+    runSaySpeed(aircraft) {
+        const speed = aircraft.speed;
+
+        return [true, `our speed is ${speed}`];
```
```
const speed = _round(aircraft.speed);

return [true, `indicating ${speed} knots`];
```

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,78 @@ export default class AircraftCommander {
+    runSayAssignedSpeed(aircraft) {
+        const speed = aircraft.target.speed;
+
+        return [true, `assigned speed ${speed}`];
```
```
if (this.speedMode !== MCP_MODE.SPEED.HOLD) {
    return [true, 'we haven't been assigned a speed'];
}

const speed = _round(aircraft.mcp.speed);

return [true, `assigned ${speed} knots`];
```

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,117 @@ export default class AircraftCommander {
+            altitudeChangeString = 'leaving ';
+        } else {
+            altitudeChangeString = 'at ';
+        }
```
No need for the else here. And stylistically, if we were going to have them say `at 6500`, we would set the variable to `at ` in the initial let declaration, thus avoiding the need for the `else` branch.

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,117 @@ export default class AircraftCommander {
+     */
+    runSayAssignedHeading(aircraft) {
+        if (aircraft.mcp.headingMode !== MCP_MODE.HEADING.HOLD) {
+            return [false, `we haven't been assigned a heading`];
```
Since there is no use of inline variable references, and this is actually a simple string, we should be using single quotes instead of back ticks. I believe lint should be showing a warning for this as well.

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,117 @@ export default class AircraftCommander {
+        if (aircraft.mcp.headingMode !== MCP_MODE.HEADING.HOLD) {
+            return [false, `we haven't been assigned a heading`];
+        }
+        const heading = _round(radiansToDegrees(aircraft.mcp.heading));
```
empty line separating all variable declarations, conditional blocks, and function calls.

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,117 @@ export default class AircraftCommander {
+     */
+    runSayAssignedSpeed(aircraft) {
+        if(this.speedMode !== MCP_MODE.SPEED.HOLD) {
+            return [false, `we haven't been assigned a speed`];
```
single quotes if no inline variable references are used

sam-irl replied:
It never occured to me to escape the '

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,117 @@ export default class AircraftCommander {
+     * @return {array} [success of operation, readback]
+     */
+    runSayAssignedSpeed(aircraft) {
+        if(this.speedMode !== MCP_MODE.SPEED.HOLD) {
```
There should be a space between "if" and the opening `(`

sam-irl replied:
This is what Java does to a person

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -509,6 +515,117 @@ export default class AircraftCommander {
+        readback.log = `assigned ${speed} knots`;
+        readback.say = `assigned ${radio_spellOut(speed)} knots}`;
+
+        return [true, `assigned speed ${speed}`];
```
this should be `[true, readback]` to make use of the readback object you created.

sam-irl commented:
@erikquinn I've addressed the feedback; however, I'm away from my main PC and I can't compile it. I tried cloning the repo and deploying it to my own Heroku instance, however, that just raised other issues and I'm no longer going there.

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -521,6 +531,120 @@ export default class AircraftCommander {
+     * @return {array} [success of operation, readback]
+     */
+    runSayAssignedSpeed(aircraft) {
+        if (this.speedMode !== MCP_MODE.SPEED.HOLD) {
```
This should be `aircraft.mcp.speedMode` in order to work correctly.

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -46,6 +50,12 @@ const COMMANDS = {
+    sayAssignedAltitude: 'runSayAssignedAltitude',
+    sayHeading: 'runSayHeading',
+    sayAssignedHeading: 'runSayAssignedHeading',
+    saySpeed: 'runSaySpeed',
```
`sayIndicatedAirspeed: 'runSayIndicatedAirspeed',`

erikquinn commented on src/assets/scripts/client/aircraft/AircraftCommander.js:
```diff
@@ -521,6 +531,120 @@ export default class AircraftCommander {
+     * @param aircraft
+     * @return {array} [success of operation, readback]
+     */
+    runSaySpeed(aircraft) {
```
`runSayIndicatedAirspeed`

erikquinn commented on src/assets/scripts/client/commandParser/commandDefinitions.js:
```diff
@@ -102,6 +102,30 @@ const ZERO_ARG_COMMANDS = {
+        validate: zeroArgumentsValidator,
+        parse: noop
+    },
+    saySpeed: {
```
sayIndicatedAirspeed

erikquinn commented on src/assets/scripts/client/commandParser/commandMap.js:
```diff
@@ -113,7 +113,13 @@ export const COMMAND_MAP = {
+    saa: 'sayAssignedAltitude',
+    sh: 'sayHeading',
+    sah: 'sayAssignedHeading',
+    ss: 'saySpeed',
```
`si: 'sayIndicatedAirspeed',`
